### PR TITLE
Revamp pod tasks

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Revamp pod tasks to use the [OpenAPI spec for PodSpec](https://github.com/kubernetes/kubernetes/tree/master/api/openapi-spec)s, increment the task_type_version accordingly and utilize the new config field in the task template to set the primary container name

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/816

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/models/task.py [hunk 1]
```diff
@@ -383,6 +394,10 @@ def to_flyte_idl(self):
             container=self.container.to_flyte_idl() if self.container else None,
             task_type_version=self.task_type_version,
         )
+        if self.config is not None:
+            for k, v in self.config.items():
+                task_template.config[k] = v
```
Comment (EngHabu) [inline flytekit/models/task.py:399]: is task_template not of type dict? are we just trying to make a copy of it?, can we not do this:
```suggestion
        config={k: v for k, v in self.config.items()} if self.config is not None else None,
    )
```

File: flytekit/models/task.py [hunk 2]
```diff
@@ -398,6 +413,7 @@ def from_flyte_idl(cls, pb2_object):
             custom=_json_format.MessageToDict(pb2_object.custom) if pb2_object else None,
             container=Container.from_flyte_idl(pb2_object.container) if pb2_object.HasField("container") else None,
             task_type_version=pb2_object.task_type_version,
+            config={k: v for k, v in _six.iteritems(pb2_object.config)},
```
Comment (EngHabu) [inline flytekit/models/task.py:416]: Is it safe to do that if config is None? or should we add 
```suggestion
            config={k: v for k, v in _six.iteritems(pb2_object.config)} if pb_obj.config is not None else None,
```

File: plugins/pod/setup.py [hunk 3]
```diff
@@ -4,7 +4,10 @@
 
 microlib_name = f"flytekitplugins-{PLUGIN_NAME}"
 
-plugin_requires = ["flytekit>=0.16.0b0,<1.0.0", "k8s-proto>=0.0.3,<1.0.0"]
+plugin_requires = [
+    "flytekit>=0.16.0b0,<1.0.0",
+    "kubernetes>=12.0.1",
```
Comment (EngHabu) [inline plugins/pod/setup.py:9]: w0oOt

File: flytekit/models/task.py (hunk 1 above)
Comment (katrogan) [reply flytekit/models/task.py]: oh this doesn't work because the config is a proto map so I can only write key by key

File: plugins/pod/flytekitplugins/pod/task.py [hunk 4]
```diff
@@ -33,7 +32,7 @@ def primary_container_name(self) -> str:
 class PodFunctionTask(PythonFunctionTask[Pod]):
     def __init__(self, task_config: Pod, task_function: Callable, **kwargs):
         super(PodFunctionTask, self).__init__(
-            task_config=task_config, task_type="sidecar", task_function=task_function, **kwargs,
+            task_config=task_config, task_type="sidecar", task_function=task_function, task_type_version=1, **kwargs,
```
Comment (kumare3) [inline plugins/pod/flytekitplugins/pod/task.py:35]: This is awesome, but as a follow up can we also create a RawPodTask? (not a PodFunctionTask), where users can just write the pod, spec and we will just run it, where they can say what is the primary container and it may look like the RawContainerTask?

Again not in this PR, maybe in the future

**Comment by kumare3** [comment]
I think this is so cool -
cc @jeevb can you also please review this?

**Comment by wild-endeavor** [comment]
Just a few questions.
1. Do the cookbook examples also need to change?
2. I looked over the IDL PR and read through the comments but I still don't understand the new config field at all.  Can you explain what it is and why it's necessary beyond the custom field?  Is it available to the plugin at execution time?  When should a plugin writer use the custom field versus the config field?
3. What line of code deals with that random question about dictionaries and protos or something that you were asking me today?

**Comment by katrogan** [comment]
@wild-endeavor:

> Do the cookbook examples also need to change?

Yes, after we cut a release this will be a breaking change. In fact now that I think about it we might need that to be the last, last beta release

> Can you explain what it is and why it's necessary beyond the custom field? Is it available to the plugin at execution time? 

Yes it's more like meta-data about the custom. In this case the primary is necessary for monitoring the execution but the custom data produced by the pod is solely the PodSpec.  Another use case could be routing for presto queries where the custom only contains the query itself.

**Comment by wild-endeavor** [comment]
So this is just entirely up to the plugin author?  They can choose to write all data to the metadata field instead right? And leave the custom blank?

File: flytekit/models/task.py [hunk 5]
```diff
@@ -370,19 +373,29 @@ def container(self):
...
-        return _core_task.TaskTemplate(
+        task_template = _core_task.TaskTemplate(
             id=self.id.to_flyte_idl(),
             type=self.type,
             metadata=self.metadata.to_flyte_idl(),
             interface=self.interface.to_flyte_idl(),
             custom=_json_format.Parse(_json.dumps(self.custom), _struct.Struct()) if self.custom else None,
             container=self.container.to_flyte_idl() if self.container else None,
             task_type_version=self.task_type_version,
+            config={k: v for k, v in self.config.items()} if self.config is not None else None,
```
Comment (wild-endeavor) [inline flytekit/models/task.py:396]: Add a comment here?  This looks like something someone might just revert to `self.config` one day.

Comment (katrogan) [reply flytekit/models/task.py:396]: I added a test that will fail if it's not correctly set

**Comment by katrogan** [comment]
@wild-endeavor 

> So this is just entirely up to the plugin author? They can choose to write all data to the metadata field instead right? And leave the custom blank?

yep

File: flytekit/models/task.py [hunk 6]
```diff
@@ -398,6 +411,7 @@ def from_flyte_idl(cls, pb2_object):
             custom=_json_format.MessageToDict(pb2_object.custom) if pb2_object else None,
             container=Container.from_flyte_idl(pb2_object.container) if pb2_object.HasField("container") else None,
             task_type_version=pb2_object.task_type_version,
+            config={k: v for k, v in _six.iteritems(pb2_object.config)} if pb2_object.config is not None else None,
```
Comment (wild-endeavor) [inline flytekit/models/task.py:414]: oh missed this sorry, can we get rid of six here and just do `.items()` like above?  we're no longer supporting python 2.

Comment (katrogan) [reply flytekit/models/task.py]: done

**Review by wild-endeavor — APPROVED** [review]
